nixos/modules/services/web-apps/node-red.nix

```diff
@@ -0,0 +1,79 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.node-red;
+in
+{
+  options.services.node-red = {
+    enable = mkEnableOption "the Node-RED service";
+
+    package = mkOption {
+      default = pkgs.nodePackages.node-red;
+      defaultText = "pkgs.nodePackages.node-red";
+      type = types.package;
+      description = "Node-RED package to use.";
+    };
+
+    configFile = mkOption {
```

Comment:
I'm left wondering if we should rename `configFile` to `settings`, change the `type` to `(pkgs.formats.json {}).type`, and then generate a `settings.js` file based on that :thinking: This would eliminate the need for the extra `define` option and make the module more configurable ... and in a declarative way.

Reply:
Maybe both? I am copying most of the layout of this module from the Cryptpad module, which itself copies from a lot of other Node services. 

Some people prefer to write their config files and specify a path to a config written in the DSL for their app, rather than writing their config entirely in Nix. What you are proposing with `settings` would make it so that it was only possible to configure the app via Nix, the config file could no longer be passed, correct?

Reply:
Some module authors choose the include both a declarative method (`settings`), as well as an override (`configFile`). If that's something you're interested in then my comment can be interpreted as renaming the `define` option to `settings`.

Be aware that the downside of doing this is that many options stop having meaning and it can cause a confusing user experience which almost certainly requires a sysadmin to read the module source code to understand certain implications.

I think in this case `settings` might be sufficient because it maps easily to a nix attribute set. Very easy to convert config.

Reply:
> Be aware that the downside of doing this is that many options stop having meaning
I've seen people say that too many options are a bad thing, yes. But it is fairly common to see a configFile option in Node modules, and other modules. This is the reason home-manager exists isn't it? This difference in opinion?

I would want both, to accommodate the people who don't want to deal with fully declarative configs and purity just yet. 

> rename configFile to settings, change the type to (pkgs.formats.json {}).type, and then generate a settings.js file based on that

Would I or would I not be able to simply pass a `path/to/config.json` if this were implemented? I want people to be able to just write and manage a config.json in their git repo, without having to interface 100% with Nix.

Reply:
`builtins.fromJSON`?

I'm just offering suggestions on this, though. Please don't take my word as final or authoritative.

Reply:
This is something that should be added by someone willing to submit he patch, since I don't feel strongly about how the config file is generated/parsed, the end result is more or less the same and it's a matter of improvement to be made to this basic module I'm providing. I think we should get it into Nix first so that it can be used, before worrying about how the configuration file is generated, although I think it's important that we provide these options sooner rather than later, I personally wouldn't have a need for it.

Reply:
Sure :+1: